vdr/didweb/web.go

```diff
@@ -0,0 +1,116 @@
+package didweb
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"io"
+	"mime"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// MethodName is the DID method name used by did:web
+const MethodName = "web"
+
+var _ types.DIDResolver = (*Resolver)(nil)
+
+// Resolver is a DID resolver for the did:web method.
+type Resolver struct {
+	HttpClient *http.Client
+}
+
+// NewResolver creates a new did:web Resolver with default TLS configuration.
+func NewResolver() *Resolver {
+	transport := http.DefaultTransport
+	if httpTransport, ok := transport.(*http.Transport); ok {
+		// Might not be http.Transport in testing
+		httpTransport = httpTransport.Clone()
+		httpTransport.TLSClientConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+		transport = httpTransport
+	}
+	return &Resolver{
+		HttpClient: &http.Client{
+			Transport: transport,
+			Timeout:   5 * time.Second,
+		},
+	}
+}
+
+// Resolve implements the DIDResolver interface.
+func (w Resolver) Resolve(id did.DID, _ *types.ResolveMetadata) (*did.Document, *types.DocumentMetadata, error) {
+	if id.Method != "web" {
+		return nil, nil, errors.New("DID is not did:web")
+	}
+
+	var baseID = id.ID
+	var path string
+	subpathIdx := strings.Index(id.ID, ":")
+	if subpathIdx == -1 {
+		path = "/.well-known/did.json"
+	} else {
+		// subpaths are encoded as / -> :
+		baseID = id.ID[:subpathIdx]
+		path = id.ID[subpathIdx:]
+		path = strings.ReplaceAll(path, ":", "/") + "/did.json"
+	}
+
+	unescapedID, err := url.PathUnescape(baseID)
```

Comment:
For supporting percent-encoded characters in the DID. You could theoretically have no "ordinary" ANSI characters in the DID's ID/path, and have all of them in hex-encoding. Also required for supporting a port number (which uses percent-encoding for `:`, to avoid having it interpreted as a subpath).

Reply:
The spec only requires restoring the port
> If the domain contains a port percent decode the colon.

encoding https://example.com/path as `did:web:example.com%2Fpath` is invalid according to `did:web` but passes these checks. I don't know how this should be handled or if it matters, just making an observation.

Reply:
I will add a check that checks if the hostname in the parsed URL equals the method-specific ID, that should cover these cases